django/contrib/messages/api.py

```diff
@@ -18,16 +18,18 @@ def add_message(request, level, message, extra_tags='', fail_silently=False):
     """
     Attempts to add a message to the request using the 'messages' app.
     """
-    if not isinstance(request, HttpRequest):
-        raise TypeError("add_message() argument must be an HttpRequest object, "
-                        "not '%s'." % request.__class__.__name__)
-    if hasattr(request, '_messages'):
+    if request and hasattr(request, '_messages'):
         return request._messages.add(level, message, extra_tags)
+
     if not fail_silently:
-        raise MessageFailure(
-            'You cannot add messages without installing '
-            'django.contrib.messages.middleware.MessageMiddleware'
-        )
+        if request and "django.contrib.messages.middleware.MessageMiddleware" not in settings.MIDDLEWARE:
+            raise MessageFailure(
+                'You cannot add messages without installing '
+                'django.contrib.messages.middleware.MessageMiddleware'
+            )
+        else:
+            raise TypeError("add_message() argument must be an HttpRequest like object, "
```

Comment:
It's not clear to me why it's valuable to change the semantics here so that this validation doesn't happen first if `fail_silently=False`.
I don't think the message must be changed to add "like".